### ISSUE TITLE
Find a way to determine the currently active keyboard configuration on Wayland

### ISSUE THREAD

**Issue by federico-terzi** [federico-terzi/espanso#921]
Currently, on Wayland we don't have a reliable way to detect the active RMLVO keyboard configuration.

For example, on X11 we could run `setxkbmap -query`, which gives us:

```
~ setxkbmap -query
rules:      evdev
model:      pc86
layout:     us
variant:    dvorak-intl
options:    terminate:ctrl_alt_bksp
```

If we manage to find a way to extract that information (Rules, Model, Layout, Variant and Options), we would greatly improve the Wayland implementation

Do you guys have any idea? :)

**Comment by phil294** [comment]
Hi @federico-terzi,

I've been building [AHK_X11](https://github.com/phil294/AHK_X11) and happen to have the same requirement as espanso. First off, I've been using your code then and again as reference for building it, and it's been great! Thank you very much for your hard work. The source is well structured, easy to read and includes helpful comments around the weirdness that some integrations pose.

I'm currently working on adding Evdev support to my program too, and want to avoid Wayland users having to specify their keyboard layout by hand. So I investigated this, and according to the official [xkbcommon quick guide](https://github.com/xkbcommon/libxkbcommon/blob/5b5ec0ee2781fb540c60f7f554787cef1e2aaa87/doc/quick-guide.md),

> If we are a Wayland client, the compositor gives us a string complete with a keymap. In this case, we can create the keymap object like this:

```c
    /* From the wl_keyboard::keymap event. */
    const char *keymap_string = <...>;
    struct xkb_keymap *keymap;

    keymap = xkb_keymap_new_from_string(ctx, keymap_string,
                                        XKB_KEYMAP_FORMAT_TEXT_V1,
                                        XKB_KEYMAP_COMPILE_NO_FLAGS);
    if (!keymap) <error>
```

There is an example that shows how to do this in complete, and it is an abysmal **700 lines** in size: https://github.com/xkbcommon/libxkbcommon/blob/master/tools/interactive-wayland.c. You can also quickly test this out by installing `libxkbcommon-tools` and running `xkbcli interactive-wayland`, which amounts to this exact example.

I haven't been able to find an alternative solution to this. It's possible I missed something, but it looks like the only solution to this is use wayland code to get notifications about current sessions, seats etc. and at some point you'll be given a keyboard layout.

Now while it's surely possible to translate this into Crystal (Rust in your case), I opted for the *much* easier solution of just altering the C code, building it statically instead and linking it to my main binary. It seems you have opted for this a couple of times for espanso too. I now do

```crystal
lib LibXKBCustom
	# this is interactive-wayland.c altered
	fun wayland_get_keymap = xkb_custom_wayland_get_keymap() : LibXKBCommon::Keymap
end
ctx = LibXKBCommon.context_new(LibXKBCommon::ContextFlags::NoFlags)
keymap = LibXKBCustom.wayland_get_keymap()
if ! keymap.null?
	# use keymap
else
	# figure out x11 keymap or use default (nil layout)
end
```
I'll be reusing most of the excellent code of @nickbclifford from https://github.com/nickbclifford/gkeybind who wrote Crystal bindings for Evdev and partially also xkbcommon.

TL;DR: This is my code: https://github.com/xkbcommon/libxkbcommon/compare/5b5ec0ee2781fb540c60f7f554787cef1e2aaa87...phil294:libxkbcommon:master and as far as I can tell from a test on a wayland system, it works great within ~100ms.

**Comment by federico-terzi** [comment]
@phil294 Thank you so much for the detailed write-up! That's extremely helpful :) 

PS: I saw AHK_X11 a while back and I was really impressed! Keep up the great work :)
